kiwi/builder/disk.py

```diff
@@ -814,16 +851,16 @@ def _build_spare_filesystem(self, device_map: Dict) -> Optional[FileSystemBase]:
                 spare_part_data_path = self.root_dir + '{0}/'.format(
                     self.spare_part_mountpoint
                 )
-            filesystem = FileSystem.new(
+            with FileSystem.new(
                 self.spare_part_fs,
                 device_map['spare'],
                 spare_part_data_path,
                 spare_part_custom_parameters
-            )
-            filesystem.create_on_device(
-                label='SPARE'
-            )
-            return filesystem
+            ) as filesystem:
+                filesystem.create_on_device(
+                    label='SPARE'
+                )
+                return filesystem
```

Comment:
The filesystem is unmounted on return. Are you sure that is the right thing to do?

Reply:
yes. I know it's not perfect but the FileSystem class supports filesystems that requires mounting to receive data and filesystems that does not e.g mksquashfs. Thus it is save to leave the context of the filesystem AND keep the instance of filesystem for later actions. I really need that and I cannot change it without further refactoring. The context manager makes sure the filesystem is not left in a mounted state. Any subsequent actions on the instance of filesystem makes sure it gets mounted when needed e.g sync_data().

Reply:
Hm, out of interest, if `sync_data()` mounts the file system again, I guess it will remain mounted for ever then, right?

Reply:
You are partially right. The point is, the ```umount``` is not a responsibility of the sync_data() method and that I also consider a design bug. The way the several sync_data() methods are called ensures that at the end (or in case of an exception) a cleanup runs which makes sure we are umounted after sync. It's the builder implementation that has to ensure this and you will find it in each kiwi/builder/... implementation. 

In the beginning I thought it will be one class that is responsible for mounting the system, sync and umount. But this was a bad design decision because over time I understand that we need mount states of the system also after sync, e.g for certain boot-loaders or to apply specific attributes on certain filesystems. At that time I decided for the shortcut, keep the mount state into the builder and let the builder decide for the right time to umount. It works but it's not super nice probably was the wrong decision back in time.

So all this would need a complete refactor of how the builders are implemented. I simply don't have the time to start this massive reconstruction and to be honest I also feel mostly left alone with the actual work. I'd like to finish the context manager move because it will increase stability but I fear along the way and as you read more of my code you will find more stuff you will not like :) The context manager move is already a big big change as it also touches everything

Reply:
@dcermak so from the perspective of the context manager initiative, can this go on ? See the build matrix it all builds. I'm fine continuing on further refactoring tasks but I also like to move forward with this one here. Thanks